### PR TITLE
Enable noUncheckedIndexedAccess for ODSP driver

### PR DESCRIPTION
**Pull request by RishhiB**
Enable noUncheckedIndexedAccess for ODSP driver
We are enabling `noUncheckedIndexedAccess` to protect us from changing a type of a record where there was a named property and reduce the risk of runtime errors by ensuring that indexed access on objects are properly checked for undefined values.

Arrays are impacted by this and must be annotated even though it is not the reason for using the setting

[AB#8216](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/8216)

### CODE REVIEW

File: packages/drivers/odsp-driver/src/odspUrlHelper.ts [hunk 1]
```diff
@@ -110,8 +110,10 @@ export async function getOdspUrlParts(url: URL): Promise<IOdspUrlParts | undefin
 			}
 		}
 
-		const driveId = joinSessionMatch[3] || joinSessionMatch[5];
-		const itemId = joinSessionMatch[4];
+		// Why are we non null asserting here
+		const driveId = joinSessionMatch[3] || joinSessionMatch[5]!;
```
Comment (RishhiB) [inline packages/drivers/odsp-driver/src/odspUrlHelper.ts:114]: its failing in CI here I need to change it to use ?? instead; https://dev.azure.com/fluidframework/public/_build/results?buildId=275487&view=logs&j=3dc8fd7e-4368-5a92-293e-d53cefc8c4b3&t=bfe27d5f-9219-57c0-0d67-1b2946a32921

File: packages/drivers/odsp-driver/src/ReadBufferUtils.ts [hunk 2]
```diff
@@ -47,7 +47,8 @@ export class ReadBuffer {
 		let length = lengthArg;
 		while (length > 0) {
 			assert(!this.eof, 0x223 /* "unexpected end of buffer" */);
-			res += this.data[this.index] * multiplier;
+			// TODO Why are we non null asserting here?
```
Comment (sonalideshpandemsft) [inline packages/drivers/odsp-driver/src/ReadBufferUtils.ts:50]: I see a lot of TODOs. how are we planning to fix them?

**Review by sonalideshpandemsft — APPROVED** [review]
Overall looks good to me. Curious about how we plan to fix all the TODOs.

**Comment by jatgarg** [comment]
@RishhiB Should we open a ADO task to remove these and then out that Task number in all these todos so that later it becomes easier to search all places where we made these changes?

**Comment by RishhiB** [comment]
@jatgarg @sonalideshpandemsft the next round after this stage we are rolling out the code changes to remove the es lint disables

**Pull request merged**